Mathlib/Algebra/Lie/Sl2.lean

```diff
@@ -0,0 +1,154 @@
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Oliver Nash
+-/
+import Mathlib.Algebra.Lie.OfAssociative
+import Mathlib.LinearAlgebra.Eigenspace.Basic
+import Mathlib.LinearAlgebra.FreeModule.StrongRankCondition
+
+/-!
+
+# The Lie algebra `sl₂` and its representations
+
+The Lie algebra `sl₂` is the unique simple Lie algebra of minimal rank, 1, and as such occupies a
+distinguished position in the general theory. This file provides some basic definitions and results
+about `sl₂`.
+
+## Main definitions:
+ * `IsSl2Triple`: a structure representing a triple of elements in a Lie algebra which satisfy the
+   standard relations for `sl₂`.
+ * `IsSl2Triple.HasPrimitiveVectorWith`: a structure representing a primitive vector in a
+   representation of a Lie algebra relative to a distinguished `sl₂` triple.
+ * `IsSl2Triple.HasPrimitiveVectorWith.exists_nat`: the eigenvalue of a primitive vector must be a
+   natural number if the representation is finite-dimensional.
+
+-/
+
+variable (R L M : Type*) [CommRing R] [LieRing L] [LieAlgebra R L]
+  [AddCommGroup M] [Module R M] [LieRingModule L M] [LieModule R L M]
+
+open LieModule
+
+variable {L} in
+/-- An `sl₂` triple within a Lie ring `L` is a triple of elements `h`, `e`, `f` obeying relations
+which ensure that the Lie subalgebra they generate is equivalent to `sl₂`. -/
+structure IsSl2Triple (h e f : L) : Prop where
+  h_ne_zero : h ≠ 0
+  lie_e_f : ⁅e, f⁆ = h
+  lie_h_e_nsmul : ⁅h, e⁆ = 2 • e
+  lie_h_f_nsmul : ⁅h, f⁆ = - (2 • f)
+
+namespace IsSl2Triple
+
+variable {L M}
+variable {h e f : L}
+
+lemma symm (ht : IsSl2Triple h e f) : IsSl2Triple (-h) f e where
+  h_ne_zero := by simpa using ht.h_ne_zero
+  lie_e_f := by rw [← neg_eq_iff_eq_neg, lie_skew, ht.lie_e_f]
+  lie_h_e_nsmul := by rw [neg_lie, neg_eq_iff_eq_neg, ht.lie_h_f_nsmul]
+  lie_h_f_nsmul := by rw [neg_lie, neg_inj, ht.lie_h_e_nsmul]
+
+@[simp] lemma symm_iff : IsSl2Triple (-h) f e ↔ IsSl2Triple h e f := by
+  suffices ∀ h e f : L, IsSl2Triple h e f → IsSl2Triple (-h) f e from
+    ⟨fun t ↦ neg_neg h ▸ this (-h) f e t, this h e f⟩
+  exact fun h e f ↦ symm
+
+lemma lie_h_e_smul (t : IsSl2Triple h e f) : ⁅h, e⁆ = (2 : R) • e :=
+  by simp [t.lie_h_e_nsmul, two_smul]
+
+lemma lie_lie_smul_f (t : IsSl2Triple h e f) : ⁅h, f⁆ = -((2 : R) • f) := by
+  simp [t.lie_h_f_nsmul, two_smul]
+
+lemma e_ne_zero (t : IsSl2Triple h e f) : e ≠ 0 := by
+  have := t.h_ne_zero
+  contrapose! this
+  simpa [this] using t.lie_e_f.symm
+
+lemma f_ne_zero (t : IsSl2Triple h e f) : f ≠ 0 := by
+  have := t.h_ne_zero
+  contrapose! this
+  simpa [this] using t.lie_e_f.symm
+
+variable {R}
+
+variable (h e f) in
+/-- Given a representation of a Lie algebra with distinguished `sl₂` triple, a vector is said to be
+primitive if it is a simultaneous eigenvector for the action of both `h`, `e`, and the eigenvalue
+for `e` is zero. -/
+structure HasPrimitiveVectorWith (m : M) (μ : R) extends IsSl2Triple h e f : Prop where
```

Comment:
After a bit more mulling over this, I'm wondering whether `extends` is the right choice. Maybe the triple should be a parameter. That seems to be how you use it in practice anyways. (And then `h e f` could possibly be implicit parameters.)

No strong opinion though.